### PR TITLE
Hifiadapterfilt

### LINKED ISSUE #7390: new module: hifiadapterfilt
Is there an existing module for this?

- [x]  I have searched for the existing module

Is there an open PR for this?

- [x] I have searched for existing PRs

Is there an open issue for this?

- [x] I have searched for existing issues

Further Information

Module for filtering adapters from pacbio hifi raw reads. I would like to include it in our long-reads nextflow assembly pipeline as a pre-processing step. 

Are you going to work on this?

- [x] I'm planning to work on this module, I added myself to the Assignees to facilitate tracking who is working on the module

### PR DESCRIPTION
**Pull request by Ge94**
## PR checklist

Closes #7390 

- [ ] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [ ] If necessary, include test data in your PR.
- [ ] Remove all TODO statements.
- [ ] Emit the `versions.yml` file.
- [ ] Follow the naming conventions.
- [ ] Follow the parameters requirements.
- [ ] Follow the input/output options guidelines.
- [ ] Add a resource `label`
- [ ] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - For modules:
    - [ ] `nf-core modules test <MODULE> --profile docker`
    - [ ] `nf-core modules test <MODULE> --profile singularity`
    - [ ] `nf-core modules test <MODULE> --profile conda`
  - For subworkflows:
    - [ ] `nf-core subworkflows test <SUBWORKFLOW> --profile docker`
    - [ ] `nf-core subworkflows test <SUBWORKFLOW> --profile singularity`
    - [ ] `nf-core subworkflows test <SUBWORKFLOW> --profile conda`

### CODE REVIEW

File: modules/nf-core/hifiadapterfilt/main.nf [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+    input:
+    tuple val(meta), path(fastq)
+
+    output:
+    tuple val(meta), path("*.filt.fastq.gz")        , emit: filt
+    tuple val(meta), path("*.contaminant.blastout") , emit: blast_search
+    tuple val(meta), path("*.stats")                , emit: stats
+    tuple val(meta), path("*.blocklist")            , emit: headers
+    path "versions.yml"                             , emit: versions
```
Comment (vagkaratzas) [inline modules/nf-core/hifiadapterfilt/main.nf:18]:
```suggestion
    tuple val(meta), path("*.filt.fastq.gz")       , emit: filt
    tuple val(meta), path("*.contaminant.blastout"), emit: blast_search
    tuple val(meta), path("*.stats")               , emit: stats
    tuple val(meta), path("*.blocklist")           , emit: headers
    path "versions.yml"                            , emit: versions
```

File: modules/nf-core/hifiadapterfilt/main.nf [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+    when:
+    task.ext.when == null || task.ext.when
+
+script:
+    def args = task.ext.args ?: ''
+    // The tool AUTOMATICALLY detects fastq files from the input folder, hence an explicit call of "fastq" is not needed.
+    def prefix = task.ext.prefix ?: "${meta.id}"
```
Comment (vagkaratzas) [inline modules/nf-core/hifiadapterfilt/main.nf:26]:
```suggestion
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    // The tool AUTOMATICALLY detects fastq files from the input folder, hence an explicit call of "fastq" is not needed.
```

**Review by vagkaratzas — CHANGES_REQUESTED** [review]
Looks good! Just some minor comments; I understand that at the moment there is only one pacbio test dataset, that unfortunately doesn't have any adapters to filter. Would it be easy/possible to upload another pacbio minimal test dataset where adapters are getting filtered out (and probably also have fewer to non empty output files)? Also, even though the conda environment and the containers describe version 3.0.0, in the versions.yml, 3.0.1 is written out. Can you investigate why the mismatch?

File: modules/nf-core/hifiadapterfilt/tests/main.nf.test.snap [hunk 3]
```diff
@@ -0,0 +1,107 @@
+{
+    "hifiadapterfilt - fastq.gz": {
+        "content": [
+            [
+                
+            ]
```
Comment (famosab) [inline modules/nf-core/hifiadapterfilt/tests/main.nf.test.snap:6]: The test produces no snaps at all - maybe this can be changed?

File: modules/nf-core/hifiadapterfilt/tests/main.nf.test [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+            }
+        }
+        then {
+            assertAll(
+                { assert process.success },
+                { assert snapshot(process.out[0][1].findAll {
+                    file(it).name == "test.filt.fastq.gz" &&
+                    file(it).name == "test.contaminant.blastout" &&
+                    file(it).name == "test.blocklist" }).match()},
+                { assert path(process.out.stats[0].get(1)).text.contains('Number of adapter contaminated ccs reads: 0 (0% of total)') },
```
Comment (famosab) [inline modules/nf-core/hifiadapterfilt/tests/main.nf.test:29]: These files seem to not be found in the snapshot?

File: modules/nf-core/hifiadapterfilt/tests/main.nf.test [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    tag "modules_nfcore"
+    tag "hifiadapterfilt"
+
+    test("hifiadapterfilt - fastq.gz") {
+        when {
+            process {
+                """
+                input[0] = [
+                    [ id:'test' ],
+                    file(params.test_data['homo_sapiens']['pacbio']['hifi'], checkIfExists: true)
```
Comment (famosab) [inline modules/nf-core/hifiadapterfilt/tests/main.nf.test:17]: can you change this to the newer path description? :) 

file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/vcf/test.vcf.gz', checkIfExists: true)

File: modules/nf-core/hifiadapterfilt/main.nf [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+    tuple val(meta), path("*.filt.fastq.gz")       , emit: filt
+    tuple val(meta), path("*.contaminant.blastout"), emit: blast_search
+    tuple val(meta), path("*.stats")               , emit: stats
+    tuple val(meta), path("*.blocklist")           , emit: headers
+    path "versions.yml"                            , emit: versions
+
+    when:
+    task.ext.when == null || task.ext.when
+
+script:
```
Comment (famosab) [inline modules/nf-core/hifiadapterfilt/main.nf:23]:
```suggestion
    script:
```

File: modules/nf-core/hifiadapterfilt/main.nf [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+    def args = task.ext.args ?: ''
+    def prefix = task.ext.prefix ?: "${meta.id}"
+    // The tool AUTOMATICALLY detects fastq files from the input folder, hence an explicit call of "fastq" is not needed.
+
```
Comment (famosab) [inline modules/nf-core/hifiadapterfilt/main.nf:27]:
```suggestion
```

**Comment by FloWuenne** [comment]
Hi @Ge94,
do you think you will be able to work on this during the upcoming hackathon or do you need help getting this PR over the finish line? 😊

**Comment by Ge94** [comment]
Hi @FloWuenne @famosab , thank you for your comments. 

I paused this PR and its failing tests, as my main issue is finding or creating a meaningful sample to launch this tool on. At the moment, those that I found have no adapters to filter, hence the output is essentially the same as the input. I tried with all hifi files I could find in nf-core/test-datasets, but none suit my case. Do you have any recommendations or documentation I could follow to generate my own sample and add it to the repo?

Thank you 😊

**Comment by famosab** [comment]
> Hi @FloWuenne @famosab , thank you for your comments.
> 
> I paused this PR and its failing tests, as my main issue is finding or creating a meaningful sample to launch this tool on. At the moment, those that I found have no adapters to filter, hence the output is essentially the same as the input. I tried with all hifi files I could find in nf-core/test-datasets, but none suit my case. Do you have any recommendations or documentation I could follow to generate my own sample and add it to the repo?
> 
> Thank you 😊

You can add new test data here: https://github.com/nf-core/test-datasets/blob/modules/README.md
For the creation i would suggest to see what is used on the tool developers side or maybe ask in the upcoming hackathon :) I am not familiar with the tool so thats the only pointer I can give.

**Comment by SPPearce** [comment]
Closing due to lack of response. Please reopen if you (or anyone else) have any plans to continue.
